File: p0-python/wasmgenerator.py
from scanner import TIMES, DIV, MOD, AND, PLUS, MINUS, OR, EQ, NE, LT, GT, LE, \
     GE, NOT, mark
from symboltable import indent, Var, Ref, Const, Type, Proc, StdProc, Int, Bool, Array, Record

def genProgStart():
    global curlev, memsize, asm
    curlev, memsize = 0, 0
    asm = ['(module',
           '(import "P0lib" "write" (func $write (param i32)))',
           '(import "P0lib" "writeln" (func $writeln))',
           '(import "P0lib" "print" (func $print (param i32)))',
           '(import "P0lib" "read" (func $read (result i32)))']

def genBool(b):
    # b is Bool
    b.size = 1; return b

def genInt(i):
    # i is Int
    i.size = 4; return i

def genRec(r):
    # r is Record
    s = 0
    for f in r.fields:
        f.offset, s = s, s + f.tp.size
    r.size = s
    return r

def genArray(a: Array):
    # a is Array
    a.size = a.length * a.base.size
    return a

def genGlobalVars(sc, start):
    global memsize
    for i in range(start, len(sc)):
        if type(sc[i]) == Var:
            if sc[i].tp in (Int, Bool):
                asm.append('(global $' + sc[i].name + ' (mut i32) i32.const 0)')
            elif type(sc[i].tp) in (Array, Record):
                sc[i].lev, sc[i].adr, memsize = -2, memsize, memsize + sc[i].tp.size
            else: mark('WASM: type?')
    
def genLocalVars(sc, start):
    for i in range(start, len(sc)):
        if type(sc[i]) == Var:
            if sc[i].tp in (Int, Bool):
                asm.append('(local $' + sc[i].name + ' i32)')
            elif type(sc[i].tp) in (Array, Record):
                mark('WASM: no local arrays, records')
            else: mark('WASM: type?')
    return None

def loadItem(x):
    if type(x) == Var:
        if x.lev == 0: asm.append('global.get $' + x.name) # global Var
        elif x.lev == curlev: asm.append('local.get $' + x.name) # local Var
        elif x.lev == -2: # memory Var
            asm.append('i32.const ' + str(x.adr))
            asm.append('i32.load')
        elif x.lev != -1: mark('WASM: var level!') # already on stack if lev == -1
    elif type(x) == Ref:
        if x.lev == -1: asm.append('i32.load')
        elif x.lev == curlev:
            asm.append('local.get $' + x.name)
            asm.append('i32.load')
        else: mark('WASM: ref level!')
    elif type(x) == Const: asm.append('i32.const ' + str(x.val))

def genVar(x):
    # x is Var, Ref
    if 0 < x.lev < curlev: mark('WASM: level!')
    if type(x) == Ref:
        y = Ref(x.tp); y.lev, y.name = x.lev, x.name
        # if type(x.tp) in (Array, Record):
        #    if x.lev > 0: y.name = x.name 
    elif type(x) == Var:
        y = Var(x.tp); y.lev, y.name = x.lev, x.name
        # if x.lev >= 0: y.name = x.name
        if x.lev == -2: y.adr = x.adr
    return y

def genConst(x):
    # x is Const
    return x

def genUnaryOp(op, x):
    loadItem(x)
    if op == MINUS:
        asm.append('i32.const -1')
        asm.append('i32.mul')
        x = Var(Int); x.lev = -1
    elif op == NOT:
        asm.append('i32.eqz')
        x = Var(Bool); x.lev = -1
    elif op == AND:
        asm.append('if (result i32)')
        x = Var(Bool); x.lev = -1
    elif op == OR:
        asm.append('if (result i32)')
        asm.append('i32.const 1')
        asm.append('else')
        x = Var(Bool); x.lev = -1
    else: mark('WASM: unary operator?')
    return x

def genBinaryOp(op, x, y):
    if op in (PLUS, MINUS, TIMES, DIV, MOD):
        loadItem(x); loadItem(y)
        asm.append('i32.add' if op == PLUS else \
                   'i32.sub' if op == MINUS else \
                   'i32.mul' if op == TIMES else \
                   'i32.div_s' if op == DIV else \
                   'i32.rem_s' if op == MOD else '?')
        x = Var(Int); x.lev = -1
    elif op == AND:
        loadItem(y) # x is already on the stack
        asm.append('else')
        asm.append('i32.const 0')
        asm.append('end')
        x = Var(Bool); x.lev = -1
    elif op == OR:
        loadItem(y) # x is already on the stack
        asm.append('end')
        x = Var(Bool); x.lev = -1
    else: assert False
    return x

def genRelation(op, x, y):
    loadItem(x); loadItem(y)
    asm.append('i32.eq' if op == EQ else \
               'i32.ne' if op == NE else \
               'i32.lt_s' if op ==  LT else \
               'i32.gt_s' if op == GT else \
               'i32.le_s' if op == LE else \
               'i32.ge_s' if op == GE else '?')
    x = Var(Bool); x.lev = -1
    return x

def genSelect(x, f):
    # x.f, assuming x.tp is Record and x is global Var, local Ref, stack Ref
    # and f is Field
    if type(x) == Var: x.adr += f.offset
    elif type(x) == Ref:
        if x.lev > 0: asm.append('local.get $' + x.name)
        asm.append('i32.const ' + str(f.offset))
        asm.append('i32.add')
        x.lev = -1
    x.tp = f.tp
    return x

def genIndex(x, y):
    # x[y], assuming x.tp is Array and x is global Var, local Ref, stack Ref
    # and y is Const, local Var, global Var, stack Var
    if type(x) == Var: # at x.adr
        if type(y) == Const: 
            x.adr += (y.val - x.tp.lower) * x.tp.base.size
            x.tp = x.tp.base
        else: # y is global Var, local Var, stack Var
            loadItem(y) # y on stack
            if x.tp.lower != 0:
                asm.append('i32.const ' + str(x.tp.lower))
                asm.append('i32.sub')
            asm.append('i32.const ' + str(x.tp.base.size))
            asm.append('i32.mul')
            asm.append('i32.const ' + str(x.adr))
            asm.append('i32.add')
            x = Ref(x.tp.base); x.lev = -1
    else: # x is local Ref, stack Ref; y is Const, global Var, local Var, stack Var
        if x.lev == curlev: loadItem(x); x.lev = -1
        if type(y) == Const:
            asm.append('i32.const ' + str((y.val - x.tp.lower) * x.tp.base.size))
            asm.append('i32.add')
        else:
            loadItem(y) # y on stack
            asm.append('i32.const ' + str(x.tp.lower))
            asm.append('i32.sub')
            asm.append('i32.const ' + str(x.tp.base.size))
            asm.append('i32.mul')
            asm.append('i32.add')
        x.tp = x.tp.base
    return x

def genAssign(x, y):
    if type(x) == Var:
        if x.lev == -2: asm.append('i32.const ' + str(x.adr))
        loadItem(y)
        if x.lev == 0: asm.append('global.set $' + x.name)
        elif x.lev == curlev: asm.append('local.set $' + x.name)
        elif x.lev == -2: asm.append('i32.store')
        else: mark('WASM: level!')
    elif type(x) == Ref:
        if x.lev == curlev: asm.append('local.get $' + x.name)
        loadItem(y)
        asm.append('i32.store')
    else: assert False

def genProgEntry(ident):
    asm.append('(func $program')

def genProgExit(x):
    asm.append(')\n(memory ' + str(memsize // 2** 16 + 1) + ')\n(start $program)\n)')
    return '\n'.join(l for l in asm)

def genProcStart(ident, fp):
    global curlev
    if curlev > 0: mark('WASM: no nested procedures')
    curlev = curlev + 1
    asm.append('(func $' + ident + ' ' + ' '.join('(param $' + e.name + ' i32)' for e in fp) + '')
    for p in fp:
        if p.tp in (Int, Bool) and type(p) == Ref:
            mark('WASM: only array and record reference parameters')
        elif type(p.tp) in (Array, Record) and type(p) == Var:
            mark('WASM: no structured value parameters')

def genProcEntry(ident, parsize, localsize):
    pass

def genProcExit(x, parsize, localsize):
    global curlev
    curlev = curlev - 1
    asm.append(')')

def genActualPara(ap, fp, n):
    if type(fp) == Ref:  #  reference parameter, assume ap is Var
        if ap.lev == -2: asm.append('i32.const ' + str(ap.adr))
        # else ap.lev == -1, on stack already
    elif type(ap) in (Var, Ref, Const): loadItem(ap)
    else: mark('unsupported parameter type')

def genCall(pr, ap):
    asm.append('call $' + pr.name)

def genRead(x):
    global curlev
    asm.append('call $read')
    if x.lev == 0: asm.append('global.set $' + x.name)
    elif x.lev == curlev: asm.append('local.set $' + x.name)
    elif x.lev == -2: asm.append('i32.store')
    else: mark('WASM: level!')
    y = Var(Int); y.lev = -1 # WTF is this supposed to do

def genWrite(x):
    loadItem(x)
    asm.append('call $write')

def genWriteln():
    asm.append('call $writeln')

def genPrint(x):
    loadItem(x)
    asm.append('call $print')

def genSeq(x, y):
    pass

def genThen(x):
    loadItem(x)
    asm.append('if')
    return x

def genIfThen(x, y):
    asm.append('end')

def genElse(x, y):
    asm.append('else')

def genIfElse(x, y, z):
    asm.append('end')

def genWhile():
    asm.append('loop')

def genDo(x):
    loadItem(x)
    asm.append('if')
    return x

def genWhileDo(t, x, y):
    asm.append('br 1')
    asm.append('end')
    asm.append('end')

File: p0-go-sequential/wasmgenerator.go
package main

import (
	"strconv"
	"strings"
)

type WasmGenerator struct {
	currentLevel int
	memorySize   int
	asm          []string
}

func (wg *WasmGenerator) GenProgStart() {
	wg.currentLevel = 0
	wg.memorySize = 0
	wg.asm = make([]string, 0)
	wg.asm = append(wg.asm, "(module")
	wg.asm = append(wg.asm, "(import \"P0lib\" \"write\" (func $write (param i32)))")
	wg.asm = append(wg.asm, "(import \"P0lib\" \"writeln\" (func $writeln))")
	wg.asm = append(wg.asm, "(import \"P0lib\" \"print\" (func $print (param i32)))")
	wg.asm = append(wg.asm, "(import \"P0lib\" \"read\" (func $read (result i32)))")
}

func (wg *WasmGenerator) GenBool(p0Type P0Type) P0Type {
	p0Type.SetSize(1)
	return p0Type
}

func (wg *WasmGenerator) GenInt(p0Type P0Type) P0Type {
	p0Type.SetSize(4)
	return p0Type
}

func (wg *WasmGenerator) GenRecord(p0Type P0Type) P0Type {
	p0record := p0Type.(*P0Record)
	var sum int = 0
	for _, param := range p0record.GetFields() {
		asVar, isVar := param.(*P0Var)
		if isVar {
			asVar.offset = sum
		}
		asRef, isRef := param.(*P0Ref)
		if isRef {
			asRef.offset = sum
		}
		sum += param.GetSize()
	}
	p0record.SetSize(sum)
	return p0record
}

func (wg *WasmGenerator) GenArray(p0Type P0Type) P0Type {
	p0array := p0Type.(*P0Array)
	p0array.SetSize(p0array.GetElementType().GetSize() * p0array.GetLength())
	return p0array
}

// GenGlobalVars creates the code for declaring the global variables at the start of the file.
// sc is a map of names to types. It represents all the global variable declarations.
// start Represents TODO:
func (wg *WasmGenerator) GenGlobalVars(sc []Entry, start int) int {
	for _, entry := range sc {
		asVar, isVar := entry.(*P0Var)
		if isVar {
			switch asVar.GetP0Type().(type) {
			case *P0Bool, *P0Int:
				(*wg).asm = append((*wg).asm, "(global $"+entry.GetName()+" (mut i32) i32.const 0)")
			case *P0Array, *P0Record:
				asVar.SetLevel(-2)
				asVar.SetAddress(wg.memorySize)
				wg.memorySize += entry.GetP0Type().GetSize()
			default:
				mark("WASM type?")
			}
		}
	}
	return 0
}

func (wg *WasmGenerator) GenLocalVars(sc []Entry, start int) int {
	for index, entry := range sc {
		asVar, isVar := entry.(*P0Var)
		if isVar && start <= index {
			switch asVar.GetP0Type().(type) {
			case *P0Int, *P0Bool:
				(*wg).asm = append((*wg).asm, "(local $"+entry.GetName()+" i32)")
			case *P0Array, *P0Record:
				mark("WASM: no local arrays, records")
			default:
				mark("WASM type?")
			}
		}
	}
	return 0
}

// The returned value is nil
func (wg *WasmGenerator) LoadItem(entry Entry) Entry {
	asVar, isVar := entry.(*P0Var)
	if isVar {
		if asVar.GetLevel() == 0 {
			wg.asm = append(wg.asm, "global.get $"+asVar.GetName())
		} else if asVar.GetLevel() == wg.currentLevel {
			wg.asm = append(wg.asm, "local.get $"+asVar.GetName())
		} else if asVar.GetLevel() == -2 {
			wg.asm = append(wg.asm, "i32.const "+strconv.Itoa(asVar.GetAddress()))
			wg.asm = append(wg.asm, "i32.load")
		}
	} else {
		asRef, isRef := entry.(*P0Ref)
		if isRef {
			if asRef.GetLevel() == -1 {
				wg.asm = append(wg.asm, "i32.load")
			} else if entry.GetLevel() == wg.currentLevel {
				wg.asm = append(wg.asm, "local.get $"+asRef.GetName())
				wg.asm = append(wg.asm, "i32.load")
			} else {
				mark("WASM: ref level!")
			}
		} else {
			asConst, isConst := entry.(*P0Const)
			if isConst {
				wg.asm = append(wg.asm, "i32.const "+strconv.Itoa(asConst.GetValue().(int)))
			}
		}
	}
	return nil
}

func (wg *WasmGenerator) GenVar(entry Entry) Entry {
	if 0 < entry.GetLevel() && entry.GetLevel() < wg.currentLevel {
		mark("WASM: level!")
	}
	var newEntry Entry
	_, isRef := entry.(*P0Ref)
	if isRef {
		newEntry = &P0Ref{entry.GetP0Type(), entry.GetName(), entry.GetLevel(), "", 0, 0}
	} else {
		asVar, isVar := entry.(*P0Var)
		if isVar {
			newEntry = &P0Var{entry.GetP0Type(), entry.GetName(), entry.GetLevel(), "", 0, 0}
			if entry.GetLevel() == -2 {
				newEntry.(*P0Var).SetAddress(asVar.GetAddress())
			}
		}
	}
	return newEntry
}

func (wg *WasmGenerator) GenConst(entry Entry) Entry {
	return entry
}

func (wg *WasmGenerator) GenUnaryOp(op int, entry Entry) Entry {
	wg.LoadItem(entry)
	switch op {
	case MINUS:
		wg.asm = append(wg.asm, "i32.const -1")
		wg.asm = append(wg.asm, "i32.mul")
		entry = &P0Var{wg.GenInt(&P0Int{}), "", -1, "", 0, 0}
	case NOT:
		wg.asm = append(wg.asm, "i32.eqz")
		entry = &P0Var{wg.GenBool(&P0Bool{}), "", -1, "", 0, 0}
	case AND:
		wg.asm = append(wg.asm, "if (result i32)")
		entry = &P0Var{wg.GenBool(&P0Bool{}), "", -1, "", 0, 0}
	case OR:
		wg.asm = append(wg.asm, "if (result i32)")
		wg.asm = append(wg.asm, "i32.const 1")
		wg.asm = append(wg.asm, "else")
		entry = &P0Var{wg.GenBool(&P0Bool{}), "", -1, "", 0, 0}
	default:
		mark("WASM: unary operator?")
	}
	return entry
}

func (wg *WasmGenerator) GenBinaryOp(op int, x, y Entry) Entry {
	switch op {
	case PLUS, MINUS, TIMES, DIV, MOD:
		wg.LoadItem(x)
		wg.LoadItem(y)
		switch op {
		case PLUS:
			wg.asm = append(wg.asm, "i32.add")
		case MINUS:
			wg.asm = append(wg.asm, "i32.sub")
		case TIMES:
			wg.asm = append(wg.asm, "i32.mul")
		case DIV:
			wg.asm = append(wg.asm, "i32.div_s")
		case MOD:
			wg.asm = append(wg.asm, "i32.rem_s")
		}
		x = &P0Var{wg.GenInt(&P0Int{}), "", -1, "", 0, 0}
	case AND:
		wg.LoadItem(y)
		wg.asm = append(wg.asm, "else")
		wg.asm = append(wg.asm, "i32.const 0")
		wg.asm = append(wg.asm, "end")
		x = &P0Var{wg.GenBool(&P0Bool{}), "", -1, "", 0, 0}
	case OR:
		// x should already be on the stack b/c magic
		wg.LoadItem(y)
		wg.asm = append(wg.asm, "end")
		x = &P0Var{wg.GenBool(&P0Bool{}), "", -1, "", 0, 0}
	default:
		panic("Unrecognized binary operator")
	}
	return x
}

func (wg *WasmGenerator) GenRelation(op int, x Entry, y Entry) Entry {
	wg.LoadItem(x)
	wg.LoadItem(y)
	switch op {
	case EQ:
		wg.asm = append(wg.asm, "i32.eq")
	case NE:
		wg.asm = append(wg.asm, "i32.ne")
	case LT:
		wg.asm = append(wg.asm, "i32.lt_s")
	case GT:
		wg.asm = append(wg.asm, "i32.gt_s")
	case LE:
		wg.asm = append(wg.asm, "i32.le_s")
	case GE:
		wg.asm = append(wg.asm, "i32.ge_s")
	default:
		panic("Unrecognized relational operator")
	}
	x = &P0Var{wg.GenBool(&P0Bool{}), "", -1, "", 0, 0}
	return x
}

// Assuming x is a Record and f is a field of x
func (wg *WasmGenerator) GenSelect(x Entry, f Entry) Entry {
	asVar, isVar := x.(*P0Var)
	if isVar {
		asVar.SetAddress(asVar.GetAddress() + f.(*P0Var).GetOffset()) // TODO: make sure that parameters are vars
		asVar.p0type = f.GetP0Type()
		return asVar
	} else {
		asRef, isRef := x.(*P0Ref)
		if isRef {
			if x.GetLevel() > 0 {
				wg.asm = append(wg.asm, "local.get $"+x.GetName())
			}
			wg.asm = append(wg.asm, "i32.const "+strconv.Itoa(f.(*P0Var).GetOffset()))
			wg.asm = append(wg.asm, "i32.add")
			asRef.SetLevel(-1)
			asRef.p0type = f.GetP0Type()
			return asRef
		}
	}
	panic("Should only call with variable of reference")
}

func (wg *WasmGenerator) GenIndex(x Entry, y Entry) Entry {
	xAsVar, xIsVar := x.(*P0Var)
	arrayType := x.GetP0Type().(*P0Array)
	if xIsVar {
		yAsConst, yIsConst := y.(*P0Const)
		if yIsConst {
			xAsVar.SetAddress(xAsVar.GetAddress() +
				(yAsConst.GetValue().(int)-arrayType.GetLowerBound())*
					arrayType.GetElementType().GetSize())
			xAsVar.p0type = xAsVar.GetP0Type().(*P0Array).GetElementType()
			return xAsVar
		}
		wg.LoadItem(y)
		if arrayType.GetLowerBound() != 0 {
			wg.asm = append(wg.asm, "i32.const "+strconv.Itoa(arrayType.GetLowerBound()))
			wg.asm = append(wg.asm, "i32.sub")
		}
		wg.asm = append(wg.asm, "i32.const "+strconv.Itoa(arrayType.GetElementType().GetSize()))
		wg.asm = append(wg.asm, "i32.mul")
		wg.asm = append(wg.asm, "i32.const "+strconv.Itoa(xAsVar.GetAddress()))
		wg.asm = append(wg.asm, "i32.add")
		x = &P0Ref{arrayType.GetElementType(), "", -1, "", 0, 0}
	} else {
		if x.GetLevel() == wg.currentLevel {
			wg.LoadItem(x)
			x.SetLevel(-1)
		}
		yAsConst, yIsConst := y.(*P0Const)
		if yIsConst {
			wg.asm = append(wg.asm, "i32.const "+
				strconv.Itoa((yAsConst.GetValue().(int)-arrayType.GetLowerBound())*arrayType.GetElementType().GetSize()))
			wg.asm = append(wg.asm, "i32.add")
		} else {
			wg.LoadItem(y)
			wg.asm = append(wg.asm, "i32.const "+strconv.Itoa(arrayType.GetLowerBound()))
			wg.asm = append(wg.asm, "i32.sub")
			wg.asm = append(wg.asm, "i32.const "+strconv.Itoa(arrayType.GetElementType().GetSize()))
			wg.asm = append(wg.asm, "i32.mul")
			wg.asm = append(wg.asm, "i32.add")
		}
		x = &P0Ref{arrayType.GetElementType(), x.GetName(), x.GetLevel(), "", 0, 0}
	}
	return x
}

func (wg *WasmGenerator) GenAssign(x, y Entry) {
	xAsVar, xIsVar := x.(*P0Var)
	xAsRef, xIsRef := x.(*P0Ref)
	if xIsVar {
		if xAsVar.GetLevel() == -2 {
			wg.asm = append(wg.asm, "i32.const "+strconv.Itoa(xAsVar.GetAddress()))
		}
		wg.LoadItem(y)
		if xAsVar.GetLevel() == 0 {
			wg.asm = append(wg.asm, "global.set $"+x.GetName())
		} else if xAsVar.GetLevel() == wg.currentLevel {
			wg.asm = append(wg.asm, "local.set $"+x.GetName())
		} else if xAsVar.GetLevel() == -2 {
			wg.asm = append(wg.asm, "i32.store")
		} else {
			mark("WASM: level!")
		}
	} else if xIsRef {
		if xAsRef.GetLevel() == wg.currentLevel {
			wg.asm = append(wg.asm, "local.get $"+x.GetName())
		}
		wg.LoadItem(y)
		wg.asm = append(wg.asm, "i32.store")
	} else {
		panic("The generator was passed code to assign to an unchangeable value")
	}
}

func (wg *WasmGenerator) GenProgEntry() { // NOTE: originally had an unused parameter `ident`
	wg.asm = append(wg.asm, "(func $program")
}

func (wg *WasmGenerator) GenProgExit() string {
	wg.asm = append(wg.asm, ")\n(memory "+strconv.Itoa(wg.memorySize/(2<<16)+1)+")\n(start $program)\n)")
	var theCode strings.Builder
	for _, line := range wg.asm {
		theCode.WriteString(line + "\n")
	}
	return theCode.String()
}

// GenProcStart generates the beginning of a procedure declaration
// ident is the name of the procedure
// fp is a slice holding the formal parameters of the procedure
func (wg *WasmGenerator) GenProcStart(ident string, fp []Entry) int {
	if wg.currentLevel > 0 {
		mark("WASM: no nested procedures")
	}
	wg.currentLevel++
	funcDecl := "(func $" + ident + " "
	for _, entry := range fp {
		_, isVar := entry.(*P0Var)
		_, isRef := entry.(*P0Ref)
		switch entry.GetP0Type().(type) {
		case *P0Int, *P0Bool:
			if isRef {
				mark("WASM: Only array and record reference parameters")
			}
		case *P0Array, *P0Record:
			if isVar {
				mark("WASM: no structured valued parameters")
			}
		}
		funcDecl += "(param $" + entry.GetName() + " i32) "
	}
	wg.asm = append(wg.asm, funcDecl)
	return 0
}

func (wg *WasmGenerator) GenProcEntry(ident string, parsize, localsize int) {
}

func (wg *WasmGenerator) GenProcExit(x Entry, parsize, localsize int) {
	wg.currentLevel--
	wg.asm = append(wg.asm, ")")
}

// ap is the actual parameter
// fp is the formal parameter
func (wg *WasmGenerator) GenActualPara(ap, fp Entry, parameterNumber int) {
	_, asRef := fp.(*P0Ref)
	if asRef {
		// Assume that ap is a Var
		if ap.GetLevel() == -2 {
			wg.asm = append(wg.asm, "i32.const "+strconv.Itoa(ap.(*P0Var).GetAddress()))
		}
	} else {
		switch ap.(type) {
		case *P0Var, *P0Ref, *P0Const:
			wg.LoadItem(ap)
		default:
			mark("Unsupported parameter type")
		}
	}
}

func (wg *WasmGenerator) GenCall(pr Entry) {
	wg.asm = append(wg.asm, "call $"+pr.GetName())
}

func (wg *WasmGenerator) GenRead(x Entry) {
	wg.asm = append(wg.asm, "call $read")
	// Dr. Sekerinski's 'hack' from the email I sent him
	y := &P0Var{wg.GenInt(&P0Int{}), "", -1, "", 0, 0}
	wg.GenAssign(x, y)
}

func (wg *WasmGenerator) GenWrite(x Entry) {
	wg.LoadItem(x)
	wg.asm = append(wg.asm, "call $write")
}

func (wg *WasmGenerator) GenWriteln() {
	wg.asm = append(wg.asm, "call $writeln")
}

func (wg *WasmGenerator) GenSeq(x, y Entry) {
}

func (wg *WasmGenerator) GenThen(x Entry) Entry {
	wg.LoadItem(x)
	wg.asm = append(wg.asm, "if")
	return x
}

func (wg *WasmGenerator) GenIfThen(x Entry) {
	wg.asm = append(wg.asm, "end")
}

func (wg *WasmGenerator) GenElse(x, y Entry) string {
	wg.asm = append(wg.asm, "else")
	return "" // Labels not used in WASM
}

func (wg *WasmGenerator) GenIfElse(string) {
	wg.asm = append(wg.asm, "end")
}

func (wg *WasmGenerator) GenWhile() string {
	wg.asm = append(wg.asm, "loop")
	return "" // Labels unused in WASM
}

func (wg *WasmGenerator) GenDo(x Entry) Entry {
	wg.LoadItem(x)
	wg.asm = append(wg.asm, "if")
	return x
}

func (wg *WasmGenerator) GenWhileDo(t string, x, y Entry) {
	wg.asm = append(wg.asm, "br 1")
	wg.asm = append(wg.asm, "end")
	wg.asm = append(wg.asm, "end")
}
